api/src/main/java/io/minio/ServerSideEncryption.java

```diff
@@ -37,209 +38,203 @@
+   *
+   * @deprecated As of release 7.0
+   */
+  @Deprecated
+  public Type getType() {
+    return this.type();
   }
 
- /**
-  * Create a new server-side-encryption object for encryption at rest (a.k.a. SSE-S3).
-  *
-  * @return an instance of ServerSideEncryption implementing SSE-S3  
-  */
-  public static ServerSideEncryption atRest() {
-    return new ServerSideEncryptionS3(); 
+  /**
+   * Set the server-side-encryption headers of this specific encryption.
+   * @param headers The metadata key-value map.
+   *
+   * @deprecated As of release 7.0
+   */
+  @Deprecated
+  public void marshal(Map<String, String> headers) {
+    headers.putAll(this.headers());
   }
 
- /**
-  * Create a new server-side-encryption object for encryption using a KMS (a.k.a. SSE-KMS).
-  * 
-  * @param keyId   specifies the customer-master-key (CMK) and must not be null.
-  * @param context is the encryption context. If the context is null no context is used.
-  *
-  * @return an instance of ServerSideEncryption implementing SSE-KMS.  
-  */
-  public static ServerSideEncryption withManagedKeys(String keyId, Map<String,String> context) 
-    throws InvalidArgumentException, UnsupportedEncodingException {
-    if (keyId == null) {
-      throw new InvalidArgumentException("The key-ID cannot be null");
-    }
-    if (context == null) {
-      return new ServerSideEncryptionKms(keyId, Optional.empty());
-    }
+  @Override
+  public boolean isDestroyed() {
+    return this.destroyed;
+  }
 
-    StringBuilder builder = new StringBuilder();
-    int i = 0;
-    builder.append('{');
-    for (Entry<String,String> entry : context.entrySet()) {
-      builder.append('"');
-      builder.append(entry.getKey());
-      builder.append('"');
-      builder.append(':');
-      builder.append('"');
-      builder.append(entry.getValue());
-      builder.append('"');
-      if (i < context.entrySet().size() - 1) {
-        builder.append(',');
-      }
+  private static boolean isCustomerKeyValid(SecretKey key) {
+    if (key == null) {
+      return false;
     }
-    builder.append('}');
-    String contextString = BaseEncoding.base64().encode(builder.toString().getBytes("UTF-8"));
-    return new ServerSideEncryptionKms(keyId, Optional.of(contextString)); 
+    return !key.isDestroyed() && key.getAlgorithm().equals("AES") && key.getEncoded().length == 32;
   }
 
   static class ServerSideEncryptionWithCustomerKey extends ServerSideEncryption {
-
     protected final SecretKey secretKey;
-    protected final MessageDigest md5;     
-   
+    protected final MessageDigest md5;
+
     public ServerSideEncryptionWithCustomerKey(SecretKey key, MessageDigest md5) {
       this.secretKey = key;
       this.md5 = md5;
     }
 
     @Override
-    public final Type getType() { 
+    public final Type type() { 
       return Type.SSE_C; 
     }
-   
+
     @Override
-    public void marshal(Map<String, String> headers) {
+    public Map<String, String> headers() {
```

Comment:
Why do we return a new HashMap instead of adding the headers to an existing one? The typical use case is to add the SSE-C headers to the HTTP header map. What do we gain here?

Reply:
Returning `Map` gives full control to users how they will be manipulated, `marshal(Map)` mutates passed values which we need to avoid for future failures.

Reply:
Not sure I understand what you mean. How is
```
Map<String, String> headers = new Map<>();
marshal(headers);
```
different from:
```
Map<String, String> headers = sse.headers();
```

apart from the situation that the the former case avoids a new map allocation when we serialize the SSE object into the HTTP headers - which is the entire purpose of this function. In the later case you crate a new map and "force" an explicit copy. So, in general both options are equally powerful just that we usually want to serialize to the HTTP header map of the request and not create a new map which we have to copy explicitly - or am I missing something.

Anyway minor.

Reply:
There is a little difference. `marshal()` mutates header map passed, there could be a possible value loss if header map contains any such header. Whereas `headers()` returns a new copy of SSE headers every time and caller has a control when/how to use returned headers.  As this is public class/method, it is not a good idea to mutate a passed header map.

Reply:
> There is a little difference. marshal() mutates header map passed, there could be a possible value loss if header map contains any such header.

That I got. I still don't see how this is different from `headers`. Sure, you make a copy but callers can do that explicitly by passing a new map when they want to avoid overwriting with `marshal`.  So getting the same behavior as `headers` like shown above. Anyway, the only purpose of this function is  to serialize the object to the HTTP request. So you have a S3 API call:
```
putObject(...) {
   req = new Request();
   req.getHeaders().add("Content-MD5", ...);
   //...   
  sse.marshal(req.getHeaders()); // With new method: req.getHeaders().addAll(sse.headers)
}
```
> As this is public class/method, it is not a good idea to mutate a passed header map.